_overviews/core/architecture-of-scala-213-collections.md

```diff
@@ -0,0 +1,1414 @@
+
+However, in all these cases, to build the right kind of collection
+you need to start with an empty collection of that kind. This is
+provided by the `empty` method, which is the last method defined in
+`PrefixMap`. This method simply returns a fresh `PrefixMap`.
+
+We'll now turn to the companion object `PrefixMap`. In fact it is not
+strictly necessary to define this companion object, as class `PrefixMap`
+can stand well on its own. The main purpose of object `PrefixMap` is to
+define some convenience factory methods. It also defines an implicit
+`Factory` for a better interoperability with other collections.
+
+The two convenience methods are `empty` and `apply`. The same methods are
+present for all other collections in Scala's collection framework so
+it makes sense to define them here, too. With the two methods, you can
+write `PrefixMap` literals like you do for any other collection:
+
+    scala> PrefixMap("hello" -> 5, "hi" -> 2)
+    res0: PrefixMap[Int] = PrefixMap(hello -> 5, hi -> 2)
+
+    scala> res0 += "foo" -> 3
+    res1: res0.type = PrefixMap(hello -> 5, hi -> 2, foo -> 3)
+
+## Summary ##
+
+To summarize, if you want to fully integrate a new collection class
+into the framework you need to pay attention to the following points:
+
+1. Decide whether the collection should be mutable or immutable.
+2. Pick the right base traits for the collection.
+3. Inherit from the right implementation trait to implement most
+   collection operations.
+4. Overload desired operations that do not return, by default, a
+   collection as specific as they could. A complete list of such
+   operations is given as an appendix.
+
+You have now seen how Scala's collections are built and how you can
+add new kinds of collections. Because of Scala's rich support for
+abstraction, each new collection type has a large number of
+methods without having to reimplement them all over again.
+
+### Acknowledgement ###
+
+These pages contain material adapted from the 2nd edition of
+[Programming in Scala](http://www.artima.com/shop/programming_in_scala) by
+Odersky, Spoon and Venners. We thank Artima for graciously agreeing to its
+publication.
+
+## Appendix: Methods to overload to support the “same result type” principle ##
+
+You want to add overloads to specialize transformation operations such that they return a more specific result type. Examples are:
+- `map`, on `StringOps`, when the mapping function returns a `Char`, should return a `String` (instead of an `IndexedSeq`),
+- `map`, on `Map`, when the mapping function returns a pair, should return a `Map` (instead of an `Iterable`),
+- `map`, on `SortedSet`, when an implicit `Ordering` is available for the resulting element type, should return a
+`SortedSet` (instead of a `Set`).
+
+The following table lists transformation operations that might return an undesirably wide type. You might want to overload
+these operations to return a more specific type.
```

Comment:
Maybe give some explanation when this is necessary - basically when the collection fixes some type parameter, like LongMap, BitSet, etc?

Reply:
Done.